crates/interpreter/src/exec.rs

```diff
@@ -1423,6 +1430,185 @@ impl<'m> Thread<'m> {
+            Some(x) => x,
+        };
+        macro_rules! convert {
+            ($T:ident, $t:ident, $s:ident) => {{
+                let atomic_val;
+                unsafe {
+                    paste::paste! {
+                        atomic_val = [< Atomic $s:upper >]::from_ptr(mem.as_mut_ptr() as *mut $s);
+                    }
+                }
+                Val::$T(atomic_val.load(Ordering::SeqCst) as $t)
+            }};
+        }
+
+        let c = match (t, n) {
+            (NumType::I32, 32) => convert!(I32, u32, u32),
+            (NumType::I64, 64) => convert!(I64, u64, u64),
+            (NumType::I32, 8) => convert!(I32, u32, u8),
+            (NumType::I32, 16) => convert!(I32, u32, u16),
+            (NumType::I64, 8) => convert!(I64, u64, u8),
+            (NumType::I64, 16) => convert!(I64, u64, u16),
+            (NumType::I64, 32) => convert!(I64, u64, u32),
+            _ => unreachable!(),
+        };
+        self.push_value(c);
+        Ok(())
+    }
+
+    #[cfg(feature = "threads")]
+    fn atomic_store(
+        &mut self, mem: &mut Memory<'m>, t: NumType, n: usize, m: MemArg,
+    ) -> Result<(), Error> {
+        let c = self.pop_value();
+        let i = self.pop_value().unwrap_i32();
+        let mem = match self.mem_slice(mem, m, i, n / 8, n as u32 / 8) {
+            None => return Err(trap()),
+            Some(x) => x,
+        };
+
+        macro_rules! convert {
+            ($s:ident, $t:ident) => {{
+                let atomic_val;
+                unsafe {
+                    paste::paste! {
+                        atomic_val = [< Atomic $t:upper >]::from_ptr(mem.as_mut_ptr() as *mut $t);
+                    }
+                }
+                paste::paste! {
+                    atomic_val.store((c.[<unwrap_ $s>]() as $t).into(),Ordering::SeqCst)
+                }
+            }};
+        }
+
+        match (t, n) {
+            (NumType::I32, 32) => convert!(i32, i32),
+            (NumType::I64, 64) => convert!(i64, i64),
+            (NumType::I32, 8) => convert!(i32, i8),
+            (NumType::I32, 16) => convert!(i32, i16),
+            (NumType::I64, 8) => convert!(i64, i8),
+            (NumType::I64, 16) => convert!(i64, i16),
+            (NumType::I64, 32) => convert!(i64, i32),
+            _ => unreachable!(),
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "threads")]
+    fn atomic_binop(
+        &mut self, op: AtomicOp, mem: &mut Memory<'m>, t: NumType, n: usize, m: MemArg,
+    ) -> Result<(), Error> {
+        let c = self.pop_value();
+        let c2 = if op == AtomicOp::Cmpxchg { self.pop_value() } else { c };
+        let i = self.pop_value().unwrap_i32();
+        let mem = match self.mem_slice(mem, m, i, n / 8, n as u32 / 8) {
+            None => return Err(trap()),
+            Some(x) => x,
+        };
+        let c = match (&op, t, n) {
+            (AtomicOp::Cmpxchg, NumType::I32, 32) => Val::I32(op.i32_binary(
```

Comment:
I suspect it might be worth splitting `AtomicOp` into `AtomicUnOp` and `AtomicBinOp` the same way non-atomic operations are. It will change a bit the syntax, but I don't think it should be too bad. Then it should even be possible to move the `fn $n(ptr: *mut $u, x: $u, y: $u)` definitions to the big `impl_op!` macro.